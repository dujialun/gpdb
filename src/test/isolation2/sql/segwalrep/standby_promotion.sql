-- Tests standby promotion triggered by FTS in 2 different scenarios.
--
-- 1st: Shut-down of master and hence unavailability of master
-- leading to standby promotion. In this case the connection between
-- master and standby is disconnected prior to promotion and
-- walreceiver doesn't exist.
--
-- 2nd: Master is alive but using fault injector simulated to not
-- respond to fts. This helps to validate fts time-out logic for
-- probes. Plus also standby promotion triggered while connection
-- between master and standby is still alive and hence walreceiver
-- also exist during promotion.

-- start_ignore
create language plpythonu;
-- end_ignore

create extension if not exists gp_inject_fault;
create or replace function pg_ctl(datadir text, command text)
returns text as $$
    import subprocess

    cmd = 'pg_ctl -D %s ' % datadir
    if command in ('stop'):
        cmd = cmd + '-w -m immediate %s' % command
    else:
        return 'Invalid command input'

    return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')
$$ language plpythonu;

create or replace function init_standby(datadir text, hostname text, port int, envport int)
returns text as $$
    import subprocess

    cmd = 'rm -rf %s ' % datadir
    subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')

    cmd = 'PGPORT=%d; gpinitstandby -a -s %s -P %d -F %s' % (envport, hostname, port, datadir)
    return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')
$$ language plpythonu;

-- set GUCs to speed-up the test
!\retcode gpconfig -c gp_fts_probe_retries -v 2;
!\retcode gpconfig -c gp_fts_probe_timeout -v 5;
!\retcode gpstop -u;

CREATE TABLE tmp_standby_promotion (datadir text, hostname text, port int, preferred_role char);
INSERT INTO tmp_standby_promotion
    SELECT datadir, hostname, port, preferred_role from gp_segment_configuration
    WHERE content = -1;

-- cache session on seg0 where master prober on it
0U: select 1;

SELECT role, preferred_role, content, mode, status FROM gp_segment_configuration;

-- stop master in order to trigger standby promotion
select pg_ctl((select datadir from gp_segment_configuration c
where c.role='p' and c.content=-1), 'stop');

-- trigger failover
0U: select gp_request_fts_probe_scan();

-- As master is down, standby has been promoted, so connect to old standby
-- port as new master
1: \c 16432;
1: select content, preferred_role, role, status, mode
from gp_segment_configuration
where content = -1;

-- init a new standby using original master data directory
-- start_ignore
1: select init_standby(
    (select datadir from tmp_standby_promotion where preferred_role = 'p'),
    (select hostname from tmp_standby_promotion where preferred_role = 'p'),
    (select port from tmp_standby_promotion where preferred_role = 'p'),
    (select port from tmp_standby_promotion where preferred_role = 'm'));
-- end_ignore
1: select content, preferred_role, role, status, mode
from gp_segment_configuration
where content = -1;

-- inject fts handler to simulate fts probe timeout
1: select gp_inject_fault_infinite('fts_handle_message', 'infinite_loop', dbid)
from gp_segment_configuration
where content = -1 and role = 'p';
1q:

-- trigger failover
0U: show gp_fts_probe_retries;
0U: show gp_fts_probe_timeout;
0U: select gp_request_fts_probe_scan();

-- trigger one more probe right away which mostly results in sending
-- promotion request again to standby, while its going through
-- promotion, which is nice condition to test as well.
0U: select gp_request_fts_probe_scan();

-- expect old master restored back to its preferred role, but standby is down
1: \c 15432;
1: select content, preferred_role, role, status, mode
from gp_segment_configuration
where content = -1;

-- set GUCs to speed-up the test
!\retcode gpconfig -r gp_fts_probe_retries;
!\retcode gpconfig -r gp_fts_probe_timeout;
!\retcode gpstop -u;

-- restore original standby
1: select pg_ctl((select datadir from tmp_standby_promotion where preferred_role = 'm'), 'stop');
-- start_ignore
1: select init_standby(
    (select datadir from tmp_standby_promotion where preferred_role = 'm'),
    (select hostname from tmp_standby_promotion where preferred_role = 'm'),
    (select port from tmp_standby_promotion where preferred_role = 'm'),
    (select port from tmp_standby_promotion where preferred_role = 'p'));
-- end_ignore

1: select content, preferred_role, role, status, mode
from gp_segment_configuration
where content = -1;

1: DROP TABLE tmp_standby_promotion;
1q:
