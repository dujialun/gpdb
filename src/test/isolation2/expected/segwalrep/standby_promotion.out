-- Tests standby promotion triggered by FTS in 2 different scenarios.
--
-- 1st: Shut-down of master and hence unavailability of master
-- leading to standby promotion. In this case the connection between
-- master and standby is disconnected prior to promotion and
-- walreceiver doesn't exist.
--
-- 2nd: Master is alive but using fault injector simulated to not
-- respond to fts. This helps to validate fts time-out logic for
-- probes. Plus also standby promotion triggered while connection
-- between master and standby is still alive and hence walreceiver
-- also exist during promotion.

-- start_ignore
create language plpythonu;
CREATE
-- end_ignore

create extension if not exists gp_inject_fault;
CREATE
create or replace function pg_ctl(datadir text, command text) returns text as $$ import subprocess 
cmd = 'pg_ctl -D %s ' % datadir if command in ('stop'): cmd = cmd + '-w -m immediate %s' % command else: return 'Invalid command input' 
return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '') $$ language plpythonu;
CREATE

create or replace function init_standby(datadir text, hostname text, port int, envport int) returns text as $$ import subprocess 
cmd = 'rm -rf %s ' % datadir subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '') 
cmd = 'PGPORT=%d; gpinitstandby -a -s %s -P %d -F %s' % (envport, hostname, port, datadir) return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '') $$ language plpythonu;
CREATE

-- set GUCs to speed-up the test
!\retcode gpconfig -c gp_fts_probe_retries -v 2;
-- start_ignore
20181212:00:32:57:032593 gpconfig:sdw:gpadmin-[INFO]:-completed successfully with parameters '-c gp_fts_probe_retries -v 2'

-- end_ignore
(exited with code 0)
!\retcode gpconfig -c gp_fts_probe_timeout -v 5;
-- start_ignore
20181212:00:32:58:032716 gpconfig:sdw:gpadmin-[INFO]:-completed successfully with parameters '-c gp_fts_probe_timeout -v 5'

-- end_ignore
(exited with code 0)
!\retcode gpstop -u;
-- start_ignore
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Starting gpstop with args: -u
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Gathering information and validating the environment...
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Obtaining Greenplum Master catalog information
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Obtaining Segment details from master...
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Greenplum Version: 'postgres (Greenplum Database) 6.0.0-alpha.0+dev.13197.g07781374ec build dev-oss'
20181212:00:32:58:000375 gpstop:sdw:gpadmin-[INFO]:-Signalling all postmaster processes to reload
. 

-- end_ignore
(exited with code 0)

CREATE TABLE tmp_standby_promotion (datadir text, hostname text, port int, preferred_role char);
CREATE
INSERT INTO tmp_standby_promotion SELECT datadir, hostname, port, preferred_role from gp_segment_configuration WHERE content = -1;
INSERT 2

-- cache session on seg0 where master prober on it
0U: select 1;
?column?
--------
1       
(1 row)

SELECT role, preferred_role, content, mode, status FROM gp_segment_configuration;
role|preferred_role|content|mode|status
----+--------------+-------+----+------
p   |p             |-1     |n   |u     
p   |p             |0      |s   |u     
m   |m             |0      |s   |u     
p   |p             |1      |s   |u     
m   |m             |1      |s   |u     
p   |p             |2      |s   |u     
m   |m             |2      |s   |u     
m   |m             |-1     |s   |u     
(8 rows)

-- stop master in order to trigger standby promotion
select pg_ctl((select datadir from gp_segment_configuration c where c.role='p' and c.content=-1), 'stop');
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

-- trigger failover
0U: select gp_request_fts_probe_scan();
gp_request_fts_probe_scan
-------------------------
t                        
(1 row)

-- As master is down, standby has been promoted, so connect to old standby
-- port as new master
1: \c 16432; 1: select content, preferred_role, role, status, mode from gp_segment_configuration where content = -1;
content|preferred_role|role|status|mode
-------+--------------+----+------+----
-1     |p             |p   |u     |s   
(1 row)

-- init a new standby using original master data directory
-- start_ignore
1: select init_standby( (select datadir from tmp_standby_promotion where preferred_role = 'p'), (select hostname from tmp_standby_promotion where preferred_role = 'p'), (select port from tmp_standby_promotion where preferred_role = 'p'), (select port from tmp_standby_promotion where preferred_role = 'm'));
init_standby                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Validating environment and parameters for standby initialization
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Checking for data directory /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/qddir/demoDataDir-1 on sdw
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:------------------------------------------------------
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master initialization parameters
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:------------------------------------------------------
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master hostname               = sdw
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master data directory         = /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/standby
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master port                   = 16432
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master hostname       = sdw
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master port           = 15432
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master data directory = /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/qddir/demoDataDir-1
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum update system catalog         = On
20181212:16:55:03:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Syncing Greenplum Database extensions to standby
20181212:16:55:04:013473 gpinitstandby:sdw:gpadmin-[INFO]:-The packages on sdw are consistent
20181212:16:55:04:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Adding standby master to catalog
20181212:16:55:04:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Database catalog updated successfully
20181212:16:55:04:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Updating pg_hbaconf file
20181212:16:55:04:013473 gpinitstandby:sdw:gpadmin-[INFO]:-pg_hbaconf files updated successfully
20181212:16:55:05:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Starting standby master
20181212:16:55:05:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Checking if standby master is running on host: sdw  in directory: /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/qddir/demoDataDir-1
20181212:16:55:06:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Cleaning up pg_hbaconf backup files
20181212:16:55:07:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Backup files of pg_hbaconf cleaned up successfully
20181212:16:55:07:013473 gpinitstandby:sdw:gpadmin-[INFO]:-Successfully created standby master on sdw

(1 row)
-- end_ignore
1: select content, preferred_role, role, status, mode from gp_segment_configuration where content = -1;
content|preferred_role|role|status|mode
-------+--------------+----+------+----
-1     |p             |p   |u     |s   
-1     |m             |m   |u     |s   
(2 rows)

-- inject fts handler to simulate fts probe timeout
1: select gp_inject_fault_infinite('fts_handle_message', 'infinite_loop', dbid) from gp_segment_configuration where content = -1 and role = 'p';
gp_inject_fault_infinite
------------------------
t                       
(1 row)
1q: ... <quitting>

-- trigger failover
0U: show gp_fts_probe_retries;
gp_fts_probe_retries
--------------------
2                   
(1 row)
0U: show gp_fts_probe_timeout;
gp_fts_probe_timeout
--------------------
5s                  
(1 row)
0U: select gp_request_fts_probe_scan();
gp_request_fts_probe_scan
-------------------------
t                        
(1 row)

-- trigger one more probe right away which mostly results in sending
-- promotion request again to standby, while its going through
-- promotion, which is nice condition to test as well.
0U: select gp_request_fts_probe_scan();
gp_request_fts_probe_scan
-------------------------
t                        
(1 row)

-- expect old master restored back to its preferred role, but standby is down
1: \c 15432; 1: select content, preferred_role, role, status, mode from gp_segment_configuration where content = -1;
content|preferred_role|role|status|mode
-------+--------------+----+------+----
-1     |p             |p   |u     |s   
(1 row)

-- set GUCs to speed-up the test
!\retcode gpconfig -r gp_fts_probe_retries;
-- start_ignore
20181212:00:33:49:001036 gpconfig:sdw:gpadmin-[INFO]:-completed successfully with parameters '-r gp_fts_probe_retries'

-- end_ignore
(exited with code 0)
!\retcode gpconfig -r gp_fts_probe_timeout;
-- start_ignore
20181212:00:33:50:001129 gpconfig:sdw:gpadmin-[INFO]:-completed successfully with parameters '-r gp_fts_probe_timeout'

-- end_ignore
(exited with code 0)
!\retcode gpstop -u;
-- start_ignore
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Starting gpstop with args: -u
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Gathering information and validating the environment...
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Obtaining Greenplum Master catalog information
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Obtaining Segment details from master...
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Greenplum Version: 'postgres (Greenplum Database) 6.0.0-alpha.0+dev.13197.g07781374ec build dev-oss'
20181212:00:33:50:001222 gpstop:sdw:gpadmin-[INFO]:-Signalling all postmaster processes to reload
. 

-- end_ignore
(exited with code 0)

-- restore original standby
1: select pg_ctl((select datadir from tmp_standby_promotion where preferred_role = 'm'), 'stop');
pg_ctl                                              
----------------------------------------------------
waiting for server to shut down done
server stopped

(1 row)
-- start_ignore
1: select init_standby( (select datadir from tmp_standby_promotion where preferred_role = 'm'), (select hostname from tmp_standby_promotion where preferred_role = 'm'), (select port from tmp_standby_promotion where preferred_role = 'm'), (select port from tmp_standby_promotion where preferred_role = 'p'));
init_standby                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Validating environment and parameters for standby initialization
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Checking for data directory /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/standby on sdw
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:------------------------------------------------------
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master initialization parameters
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:------------------------------------------------------
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master hostname               = sdw
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master data directory         = /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/qddir/demoDataDir-1
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum master port                   = 15432
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master hostname       = sdw
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master port           = 16432
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum standby master data directory = /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/standby
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Greenplum update system catalog         = On
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Syncing Greenplum Database extensions to standby
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-The packages on sdw are consistent
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Adding standby master to catalog
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Database catalog updated successfully
20181212:16:55:50:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Updating pg_hbaconf file
20181212:16:55:51:014164 gpinitstandby:sdw:gpadmin-[INFO]:-pg_hbaconf files updated successfully
20181212:16:55:52:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Starting standby master
20181212:16:55:52:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Checking if standby master is running on host: sdw  in directory: /home/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/standby
20181212:16:55:53:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Cleaning up pg_hbaconf backup files
20181212:16:55:53:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Backup files of pg_hbaconf cleaned up successfully
20181212:16:55:53:014164 gpinitstandby:sdw:gpadmin-[INFO]:-Successfully created standby master on sdw

(1 row)
-- end_ignore

1: select content, preferred_role, role, status, mode from gp_segment_configuration where content = -1;
content|preferred_role|role|status|mode
-------+--------------+----+------+----
-1     |p             |p   |u     |s   
-1     |m             |m   |u     |s   
(2 rows)

1: DROP TABLE tmp_standby_promotion;
DROP
1q: ... <quitting>
